# Missouri Sex Offender Registry

Data acquisition, documentation, carpentry, geocoding, and database loading for Missouri Sex Offender Registry (MSOR) and supporting info.   

In [63]:
# IMPORTS
import geopandas as gpd
import pandas as pd

# import os
# import urllib.request
# import requests
# import shutil
# from pathlib import Path
# from zipfile import ZipFile

import matplotlib.pyplot as plt
from matplotlib import pyplot

import folium

from shapely.geometry import Point, Polygon

# from geopandas.tools import overlay

from geopy.geocoders import Nominatim # for geocoding

import random # for obscuring sex offender names

In [2]:
# we need GeoAlchemy2 to run the geodataframe to_postgis method later

In [4]:
pip install GeoAlchemy2


Note: you may need to restart the kernel to use updated packages.


In [5]:
# a few more imports specfic to the database process
import geoalchemy2 
import getpass

import psycopg2
import numpy
from psycopg2.extensions import adapt, register_adapter, AsIs

from sqlalchemy import create_engine


In [6]:
# get user password for connecting to the db
mypasswd = getpass.getpass()

········


In [7]:
# set up db connection
conn = psycopg2.connect(database = 'cappsds_psmd39', 
                              user = 'psmd39', 
                              host = 'pgsql.dsa.lan',
                              password = mypasswd)


In [8]:
# establish cursor and read the existing tables
cursor = conn.cursor()

cursor.execute("""SELECT relname FROM pg_class WHERE relkind='r'
                  AND relname !~ '^(pg_|sql_)';""") # "rel" is short for relation.

tables = [i[0] for i in cursor.fetchall()] # A list() of tables.
tables.sort()
tables


['country_borders',
 'gadm_admin_borders',
 'geonames_feature',
 'msorfailedgeocoding',
 'spatial_ref_sys',
 'stlchildcare',
 'stlnonrestrictedresidential',
 'stlnonrestrictedresparcels',
 'stlpubschools',
 'stlpvtschools',
 'stlresparcels',
 'stlrestrictedflat',
 'stlsexoffenders',
 'stlzoning']

## Failed geocoding
There were a lot of sex offender registry entries that failed geocoding. Possible solutions:
- Look for trends in the failed entries and work to resolve
- Try another geocoder  

### Get the entries that failed geocoding out of the database
In the prior notebook, we stored all these records in a dedicated table for easy access.

In [9]:
# query the table and read data into a df 
sql = "select * from msorfailedgeocoding;"
msor_nogeo = pd.read_sql_query(sql, conn)
print(msor_nogeo.shape)
msor_nogeo.head()

(1334, 14)


index                  name                    address         city  st  \
0     10       ABDI, IBRAHIM A     3764 CHIPPEWA ST APT 8  SAINT LOUIS  MO   
1     18  ABERNATHY, RANDELL L   3866 S SPRING AVE APT 1S  SAINT LOUIS  MO   
2     40    ACKLEY, CLIFFORD D        3329 LAWN AVE APT 4  SAINT LOUIS  MO   
3     41    ACKLEY, CLIFFORD D        3329 LAWN AVE APT 4  SAINT LOUIS  MO   
4     99   ADAMS, QUINDARRYL L  4133 CLEVELAND AVE APT 1W  SAINT LOUIS  MO   

     zip         county                   offense  count compliant  tier  \
0  63116  ST LOUIS CITY     SEXUAL MISCONDUCT-3RD      1         Y     1   
1  63116  ST LOUIS CITY      AGG CRIM SEX ASSAULT      2         Y     3   
2  63139  ST LOUIS CITY                      RAPE      1         Y     3   
3  63139  ST LOUIS CITY  STATUTORY RAPE-2ND DEGRE      1         Y     3   
4  63110  ST LOUIS CITY            SEXUAL BATTERY      1         Y     1   

  date_of_birth                              full_address geocode  
0    1981-09-08     3764 CHIPPEWA ST APT 8,SAINT LOUIS,MO    None  
1    1969-07-30   3866 S SPRING AVE APT 1S,SAINT LOUIS,MO    None  
2    1964-10-25        3329 LAWN AVE APT 4,SAINT LOUIS,MO    None  
3    1964-10-25        3329 LAWN AVE APT 4,SAINT LOUIS,MO    None  
4    1990-10-30  4133 CLEVELAND AVE APT 1W,SAINT LOUIS,MO    None

### Set up the geocoder

In [10]:
# set up the geocoder
geolocator = Nominatim(timeout=10, user_agent = "myGeolocator")

In [11]:
# test out the geocoder with a single address
location = geolocator.geocode('120 CATALAN,ST LOUIS,MO')
print(location)
print((location.latitude, location.longitude))

St. Louis Skatium, 120, East Catalan Street, Patch, Saint Louis, Missouri, 63111, United States
(38.5396446, -90.26550765004728)


### Remove some of the substrings that cause the geocoder to fail

In [17]:
# set up a list containing the string elements we want to remove
to_remove = [' FL ',' APT',' NBR',' RM',' UNIT',' DEPT',' REAR']
tot_ct = 0 # initialize a counter

# copy the existing addresses to a new column to initialize the target of the 'for' loop
msor_nogeo['new_address'] = msor_nogeo['address']
print("Dataframe has",len(msor_nogeo),"entries")

# loop through all the elements in the list, removing each from the address
for i in to_remove:
    # split() outputs a list of two elements: the part of the address before the match string [i] and the part after
    # ...we only care about the part before
    msor_nogeo['split'] = msor_nogeo['new_address'].str.split(i)
    # convert those list items into two columns in a new (temp) df
    # then store the usable column back in the original df
    address_split = pd.DataFrame(msor_nogeo['split'].to_list(), columns=['keep', 'trash'])
    # count how many items we modified
    loop_ct = address_split['trash'].notnull().sum()
    # keep a running total of the modifications we've made
    tot_ct = tot_ct + loop_ct
    # overwrite the "new_address" with the updated value. this can then be used in subsequent loops for new matches.
    msor_nogeo['new_address'] = address_split['keep']
    print('Removed "',i,'" from address',' (',loop_ct,' entries)',sep='')

# drop the split column since we don't need it anymore
msor_nogeo.drop('split', axis=1, inplace=True)

print(tot_ct,'total modifications')

msor_nogeo[['zip','address','full_address','new_address']].head()


Dataframe has 1334 entries
Removed " FL " from address (173 entries)
Removed " APT" from address (992 entries)
Removed " NBR" from address (3 entries)
Removed " RM" from address (53 entries)
Removed " UNIT" from address (17 entries)
Removed " DEPT" from address (1 entries)
Removed " REAR" from address (2 entries)
1241 total modifications


zip                    address                              full_address  \
0  63116     3764 CHIPPEWA ST APT 8     3764 CHIPPEWA ST APT 8,SAINT LOUIS,MO   
1  63116   3866 S SPRING AVE APT 1S   3866 S SPRING AVE APT 1S,SAINT LOUIS,MO   
2  63139        3329 LAWN AVE APT 4        3329 LAWN AVE APT 4,SAINT LOUIS,MO   
3  63139        3329 LAWN AVE APT 4        3329 LAWN AVE APT 4,SAINT LOUIS,MO   
4  63110  4133 CLEVELAND AVE APT 1W  4133 CLEVELAND AVE APT 1W,SAINT LOUIS,MO   

          new_address  
0    3764 CHIPPEWA ST  
1   3866 S SPRING AVE  
2       3329 LAWN AVE  
3       3329 LAWN AVE  
4  4133 CLEVELAND AVE

In [19]:
# work up new addresses that are geocoder-compatible
msor_nogeo['new_address'] = msor_nogeo['new_address'] + "," + msor_nogeo['city'] + "," + msor_nogeo['st']


In [20]:
# count no geocodes (isnull=='True') BEFORE sending to geocoder
msor_nogeo['geocode'].isnull().value_counts()
# msor_nogeo['geocode'].count() # count the number of non-None - i.e. successful geocode - elements

True    1334
Name: geocode, dtype: int64

In [21]:
# send the updated addresses back to the geocoder
msor_nogeo['geocode'] = msor_nogeo.new_address.apply(geolocator.geocode)


In [23]:
# count no geocodes AFTER sending to geocoder
msor_nogeo['geocode'].isnull().value_counts()

False    1222
True      112
Name: geocode, dtype: int64

### How are we doing?
By looking at the amount of null/"None" elements in the 'geocode' columnn before and after running our updated addresses through the geocoder, we can see a significant improvement in our outcome. We've fixed over 90% of the failed entries! 

Let's visualize the results so far, then save this work by pushing the now-successful entries into the PostGIS database.

In [25]:
# downselect our 'msor_nogeo' gdf to only the items that now have geocodes
# remove rows that do not have location data
msor_nogeo_fixed = msor_nogeo.dropna().copy()

# find all rows where the geocode still did not populate
# save them in a new df so we can examine them later
msor_nogeo_v2 = msor_nogeo[msor_nogeo['geocode'].isna()].copy()


In [30]:
# set up the gdf to visualize the results
# get the latitude and longitude values from the geodata column and put them in their own columns for easier plotting
msor_nogeo_fixed['lat'] = [g.latitude for g in msor_nogeo_fixed.geocode]
msor_nogeo_fixed['long'] = [g.longitude for g in msor_nogeo_fixed.geocode]


#### Render a map that shows all the entries we recovered!

In [31]:
# create a base map centered on St. Louis
map_sexoffenders = folium.Map(
    location=[38.627003, -90.3],
    tiles='cartodbpositron',
    zoom_start=11,
)

# add a marker for each childcare facility
# label each facility with its name
for i in range(0,len(msor_nogeo_fixed)):
   folium.Marker(
      location=[msor_nogeo_fixed.iloc[i]['lat'], msor_nogeo_fixed.iloc[i]['long']],
      popup=msor_nogeo_fixed.iloc[i]['offense']
   ).add_to(map_sexoffenders)

# display the map
map_sexoffenders

#### Append these results to the existing table

In [33]:
# check out the form of the existing 'stlsexoffenders' table 
# we will need to match this structure in order to successfully append our new entries

# query the table and read data into a df 
sql = "select * from stlsexoffenders LIMIT 10;"
msor_table_sample = pd.read_sql_query(sql, conn)
print(msor_table_sample.shape)
print(msor_table_sample.dtypes)
msor_table_sample.head(2)


(10, 14)
randomid       int64
name          object
address       object
city          object
st            object
zip            int64
county        object
offense       object
count          int64
compliant     object
tier           int64
lat          float64
long         float64
geometry      object
dtype: object


randomid              name        address         city  st    zip  \
0     11929  ABBOTT, STEVEN R  1621 N 1ST ST  SAINT LOUIS  MO  63102   
1     14856  ABBOTT, STEVEN R  1621 N 1ST ST  SAINT LOUIS  MO  63102   

          county                  offense  count compliant  tier        lat  \
0  ST LOUIS CITY  CHILD MOLEST-1ST DEGREE      1         Y     3  38.641888   
1  ST LOUIS CITY    SEXUAL MISCONDUCT-1ST      1         Y     3  38.641888   

        long                                           geometry  
0 -90.183579  0101000020E6100000A36A2EC0BF8B56C0F1BB37612952...  
1 -90.183579  0101000020E6100000A36A2EC0BF8B56C0F1BB37612952...

In [34]:
# compare this to the same info from the df we're been working on
print(msor_nogeo_fixed.shape)
print(msor_nogeo_fixed.dtypes)
msor_nogeo_fixed.head(2)

(1222, 17)
index                     int64
name                     object
address                  object
city                     object
st                       object
zip                       int64
county                   object
offense                  object
count                     int64
compliant                object
tier                      int64
date_of_birth    datetime64[ns]
full_address             object
geocode                  object
new_address              object
lat                     float64
long                    float64
dtype: object


index                  name                   address         city  st  \
0     10       ABDI, IBRAHIM A    3764 CHIPPEWA ST APT 8  SAINT LOUIS  MO   
1     18  ABERNATHY, RANDELL L  3866 S SPRING AVE APT 1S  SAINT LOUIS  MO   

     zip         county                offense  count compliant  tier  \
0  63116  ST LOUIS CITY  SEXUAL MISCONDUCT-3RD      1         Y     1   
1  63116  ST LOUIS CITY   AGG CRIM SEX ASSAULT      2         Y     3   

  date_of_birth                             full_address  \
0    1981-09-08    3764 CHIPPEWA ST APT 8,SAINT LOUIS,MO   
1    1969-07-30  3866 S SPRING AVE APT 1S,SAINT LOUIS,MO   

                                             geocode  \
0  (3764, Chippewa Street, Dutchtown, Saint Louis...   
1  (3866, South Spring Avenue, Dutchtown, Saint L...   

                        new_address        lat       long  
0   3764 CHIPPEWA ST,SAINT LOUIS,MO  38.588789 -90.249231  
1  3866 S SPRING AVE,SAINT LOUIS,MO  38.587078 -90.248553

Comparing the outputs above, we need to...  
1. **Add** a randomid  
2. **Remove** index, date_of_birth, full_address, geocode, new_address
3. **Convert** our dataframe into a GEOdataframe

In [59]:
# 1. add a random ID as we did in the initial run 

# find the maximum randomid value from the existing data and set this value to a variable
# this will set the minimum for our new random values
sql = "select max(randomid) from stlsexoffenders;"
max_randomid = pd.read_sql_query(sql, conn)
max_randomid = max_randomid.loc[max_randomid.index,'max'].iat[0]

# get a count of how many entries we have in the dataframe
offend_ct = len(msor_nogeo_fixed.index)

# generate a random ID for each entry
# here we're taking the count of entries, multiplying it by 10, generating a range of that many sequential numbers,
#     then selecting a random sample of those values to assign to our dataframe

# we can use `insert` to place the new column in the first position
msor_nogeo_fixed.insert(0, 'randomid', random.sample(range(max_randomid+1,max_randomid+offend_ct*10), offend_ct))


In [68]:
# before we start dropping columns, copy the dataframe just in case
msor_db = msor_nogeo_fixed.copy()

In [69]:
# 2. drop columns that we don't need
msor_db.drop(['index','date_of_birth','full_address','geocode','new_address'], inplace=True, axis=1)

In [70]:
# 3. convert dataframe into a geodataframe in order for it to work correctly with PostGIS

# create the 'geometry' column for the geodataframe
geometry = [Point(xy) for xy in zip(msor_db['long'], msor_db['lat'])]
# generate the geodataframe using the msor df + the geometry info
# set the CRS (in degrees) as part of this process
msor_db = gpd.GeoDataFrame(msor_db, geometry = geometry, crs=4326) 


In [84]:
# load the data!

# Set up database connection engine
# FORMAT: engine = create_engine('postgresql://user:password@host:5432/')
engine = create_engine('postgresql://psmd39:Mizzou23?@pgsql.dsa.lan:5432/cappsds_psmd39', echo=False)

# GeoDataFrame to PostGIS
msor_db.to_postgis(
    con=engine,
    name="stlsexoffenders",
    if_exists='append' # note that we are APPENDING this new info to the existing table
)


In [88]:
engine.dispose() 

**Our database table now contains the original entries plus the newly-fixed entries.**  
However, we still have a few items that have not been successfully geocoded. Let's store those in a NEW table so we can work on them later without running through this entire process again.

In [89]:
# Set up database connection engine
engine = create_engine('postgresql://psmd39:Mizzou23?@pgsql.dsa.lan:5432/cappsds_psmd39', echo=False)

# DataFrame to PostgreSQL
msor_nogeo_v2.to_sql(
    con=engine,
    name="msorfailedgeocodingv2",
    if_exists='replace'
)

## Failed geocoding redux
Can we get the amount of geocoding failures even closer to zero? Potential trouble areas in the remaining data:

- Cardinal direction letters e.g. "N","S","E","W"
- Road suffixes e.g. "AVE","RD","BLVD"

**NOTE:** We are doing all these in a second pass (vs. rolling into the above work) because making too many changes at once has an adverse effect on many entries. That is, removing things like " FL" (up above) was enough to get those items to geocode, but removing *more* info like "AVE FL" could cause those same items to fail. Thus, this iterative approach is needed.

In [137]:
# query the table and read data into a df 
sql = "select * from msorfailedgeocodingv2;"
msor_nogeo_redux = pd.read_sql_query(sql, conn)
print(msor_nogeo_redux.shape)


(112, 16)


In [138]:
msor_nogeo_redux[['zip','address','full_address','new_address']].head()


zip                    address                           full_address  \
0  63111  120 W CATALAN AVE APT 201  120 W CATALAN AVE APT 201,ST LOUIS,MO   
1  63111  120 W CATALAN AVE APT 201  120 W CATALAN AVE APT 201,ST LOUIS,MO   
2  63111  120 W CATALAN AVE APT 201  120 W CATALAN AVE APT 201,ST LOUIS,MO   
3  63111  120 W CATALAN AVE APT 201  120 W CATALAN AVE APT 201,ST LOUIS,MO   
4  63144              2631 SALEM RD           2631 SALEM RD,SAINT LOUIS,MO   

                     new_address  
0  120 W CATALAN AVE,ST LOUIS,MO  
1  120 W CATALAN AVE,ST LOUIS,MO  
2  120 W CATALAN AVE,ST LOUIS,MO  
3  120 W CATALAN AVE,ST LOUIS,MO  
4   2631 SALEM RD,SAINT LOUIS,MO

In [139]:
# we will be working primarily off of the 'new_address' column in order to benefit from the earlier modifications

# now remove some of the elements that trip up the geocoder, using the comma to avoid unwanted replacements elsewhere
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' RD,',',')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' AVE,',',')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' DR,',',')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' ST,',',')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' BLVD,',',')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' LN,',',')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' TRAK,',',')

# remove all cardinal directions
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' N ',' ')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' S ',' ')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' E ',' ')
msor_nogeo_redux['new_address'] = msor_nogeo_redux['new_address'].str.replace(' W ',' ')


In [140]:
# count no geocodes (isnull=='True') BEFORE sending to geocoder
msor_nogeo_redux['geocode'].isnull().value_counts()
# msor_nogeo['geocode'].count() # count the number of non-None - i.e. successful geocode - elements

True    112
Name: geocode, dtype: int64

In [141]:
# send the updated addresses back to the geocoder
msor_nogeo_redux['geocode'] = msor_nogeo_redux.new_address.apply(geolocator.geocode)


In [142]:
# count no geocodes AFTER sending to geocoder
msor_nogeo_redux['geocode'].isnull().value_counts()

False    63
True     49
Name: geocode, dtype: int64

We were able to recover over half (63 of 112) of the remaining items! Let's get them into the database and wrap up this work.

In [144]:
# downselect our 'msor_nogeo' gdf to only the items that now have geocodes
# remove rows that do not have location data
msor_nogeo_fixed_redux = msor_nogeo_redux.dropna().copy()

# find all rows where the geocode still did not populate
# save them in a new df so we can examine them later
msor_nogeo_v3 = msor_nogeo_redux[msor_nogeo_redux['geocode'].isna()].copy()


In [145]:
# set up the gdf to visualize the results
# get the latitude and longitude values from the geodata column and put them in their own columns for easier plotting
msor_nogeo_fixed_redux['lat'] = [g.latitude for g in msor_nogeo_fixed_redux.geocode]
msor_nogeo_fixed_redux['long'] = [g.longitude for g in msor_nogeo_fixed_redux.geocode]


#### Render a map that shows all the entries we recovered!

In [146]:
# create a base map centered on St. Louis
map_sexoffenders_redux = folium.Map(
    location=[38.627003, -90.3],
    tiles='cartodbpositron',
    zoom_start=11,
)

# add a marker for each childcare facility
# label each facility with its name
for i in range(0,len(msor_nogeo_fixed_redux)):
   folium.Marker(
      location=[msor_nogeo_fixed_redux.iloc[i]['lat'], msor_nogeo_fixed_redux.iloc[i]['long']],
      popup=msor_nogeo_fixed_redux.iloc[i]['offense']
   ).add_to(map_sexoffenders_redux)

# display the map
map_sexoffenders_redux

#### Append these results to the existing table

In [148]:
# compare this to the same info from the df we're been working on
print(msor_nogeo_fixed_redux.shape)
msor_nogeo_fixed_redux.head(2)

(63, 18)


level_0  index                name                    address      city  \
0        9    212  ALDRIDGE, SAMUEL A  120 W CATALAN AVE APT 201  ST LOUIS   
1       10    213  ALDRIDGE, SAMUEL A  120 W CATALAN AVE APT 201  ST LOUIS   

   st    zip         county                   offense  count compliant  tier  \
0  MO  63111  ST LOUIS CITY              ATTEMPT RAPE      1         Y     3   
1  MO  63111  ST LOUIS CITY  CHLD MOLST-2ND DEG-INJRY      1         Y     3   

  date_of_birth                           full_address  \
0    1964-10-22  120 W CATALAN AVE APT 201,ST LOUIS,MO   
1    1964-10-22  120 W CATALAN AVE APT 201,ST LOUIS,MO   

                                             geocode              new_address  \
0  (St. Louis Skatium, 120, East Catalan Street, ...  120 CATALAN,ST LOUIS,MO   
1  (St. Louis Skatium, 120, East Catalan Street, ...  120 CATALAN,ST LOUIS,MO   

         lat       long  
0  38.539645 -90.265508  
1  38.539645 -90.265508

Comparing the outputs above, we need to...  
1. **Add** a randomid  
2. **Remove** index, date_of_birth, full_address, geocode, new_address
3. **Convert** our dataframe into a GEOdataframe

In [149]:
# 1. add a random ID as we did in the initial run 

# find the maximum randomid value from the existing data and set this value to a variable
# this will set the minimum for our new random values
sql = "select max(randomid) from stlsexoffenders;"
max_randomid = pd.read_sql_query(sql, conn)
max_randomid = max_randomid.loc[max_randomid.index,'max'].iat[0]

# get a count of how many entries we have in the dataframe
offend_ct = len(msor_nogeo_fixed_redux.index)

# generate a random ID for each entry
# here we're taking the count of entries, multiplying it by 10, generating a range of that many sequential numbers,
#     then selecting a random sample of those values to assign to our dataframe

# we can use `insert` to place the new column in the first position
msor_nogeo_fixed_redux.insert(0, 'randomid', random.sample(range(max_randomid+1,max_randomid+offend_ct*10), offend_ct))


In [155]:
# before we start dropping columns, copy the dataframe just in case
msor_db = msor_nogeo_fixed_redux.copy()


In [156]:
# 2. drop columns that we don't need
msor_db.drop(['index','level_0','date_of_birth','full_address','geocode','new_address'], inplace=True, axis=1)

In [157]:
# 3. convert dataframe into a geodataframe in order for it to work correctly with PostGIS

# create the 'geometry' column for the geodataframe
geometry = [Point(xy) for xy in zip(msor_db['long'], msor_db['lat'])]
# generate the geodataframe using the msor df + the geometry info
# set the CRS (in degrees) as part of this process
msor_db = gpd.GeoDataFrame(msor_db, geometry = geometry, crs=4326) 


In [160]:
# load the data!

# Set up database connection engine
# FORMAT: engine = create_engine('postgresql://user:password@host:5432/')
engine = create_engine('postgresql://psmd39:Mizzou23?@pgsql.dsa.lan:5432/cappsds_psmd39', echo=False)

# GeoDataFrame to PostGIS
msor_db.to_postgis(
    con=engine,
    name="stlsexoffenders",
    if_exists='append' # note that we are APPENDING this new info to the existing table
)


In [162]:
#close connection to the db
conn.close()
engine.dispose()

# Summary

We've successfull added as many of the sex offender locations to our PostGIS database as is reasonable. The remaining few (49 out of approximately 3,500) should not have a meaningful impact on our analysis. Some of these remaining items, such as "complaint/pending registration" and "homeless" aren't actual addresses and thus will never geocode. It's time to move on to additional work.